### ISSUE TITLE
[C++] Compiler may generate temp allocations within a loop

### ISSUE THREAD

**Issue by jgranick** [HaxeFoundation/haxe#6924]
Certain use-cases may cause the compiler to generate temporary variables within a loop. On C++, especially, this causes a lot of garbage. This is affecting inside of very large loops where we process a lot of data.

Here's some example code that reproduces it:

```haxe
class Main {
	
	public static function main () {
		
		var test = new TestClass ();
		
		var total = 0.0;
		
		for (i in 0...10000) {
			
			total += add (test.getRandom (), test.getRandom ());
			
		}
		
		trace (total);
		
	}
	
	private static function add (a:Float, b:Float):Float {
		
		return a + b;
		
	}
	
}


class TestClass {
	
	public function new () {}
	
	public function getRandom ():Float {
		
		return Math.random ();
		
	}
	
}
```

The relevant C++ generated code looks like:

```cpp
HXLINE(   5)		 ::TestClass test =  ::TestClass_obj::__alloc( HX_CTX );
HXLINE(   7)		Float total = ((Float)0.0);
HXLINE(   9)		{
HXLINE(   9)			int _g = (int)0;
HXDLIN(   9)			while((_g < (int)10000)){
HXLINE(   9)				_g = (_g + (int)1);
HXDLIN(   9)				int i = (_g - (int)1);
HXLINE(  11)				Float total1 = test->getRandom();
HXDLIN(  11)				total = (total + ::Main_obj::add(total1,test->getRandom()));
            			}
            		}
```

Easier said than done, but putting `Float total1` outside of the loop would result in less garbage. In our case, we see this occur dozens of times within a large loop, leading to collections of up to a million objects.

**Comment by nadako** [comment]
I don't see where allocations are made in this snippet - `Float` is just a `double`.

**Comment by PXshadow** [comment]
It doesn't matter the type, if you can define the variable outside of the while loop you aren't allocating for a new variable as well as the eventual garbage collection every iteration, the resources would be greatly reduced compared to the above example from @jgranick

**Comment by ibilon** [comment]
@nadako means an actual c++ double https://github.com/HaxeFoundation/hxcpp/blob/master/include/hxcpp.h#L200 which aren't part of the garbage collection or anything.

and 99% sure that all c++ compiler optimize such a case.

**Comment by ousado** [comment]
Here is a [flamegraph](http://ousado.github.io/hxcpp_flamegraph/test.hx.svg?x=y) from your example - besides generating random numbers, there's absolutely nothing going on in that code.

**Comment by ousado** [comment]
Also, Haxe has strict left to right evaluation semantics, C++ doesn't, hence that temporary var that ensures the correct evaluation order.

**Comment by ibilon** [comment]
@Gama11 thinks it might be the float wrapped in a `::Dynamic` when using `openfl.Vector` that you see getting collected.

**Comment by hughsando** [comment]
Yes - it looks like this code is ok, and a more complex example might be required.
It would need the "alloc" call, or an implicit or explicit Dynamic constructor in order to create an allocation.
